File: Ruby/lib/decorators/time_sensitive.rb
# frozen_string_literal: true

require 'delegate'

module Decorators
  # Increases an item's quality increasingly rapidly until quality becomes zero
  class TimeSensitive < SimpleDelegator
    MAXIMUM_QUALITY = 50

    alias item __getobj__

    def tick
      item.days_remaining -= 1

      return item.quality = 0 if days_remaining < 0

      item.quality += calculate_increase
      item.quality = [quality, MAXIMUM_QUALITY].min
    end

    private def calculate_increase
      increase = 1
      increase += 1 if days_remaining < 10
      increase += 1 if days_remaining < 5

      increase
    end
  end
end

File: JavaScript/src/decorators/TimeSensitive.js
import ItemDecorator from './ItemDecorator';

// Increases an item's quality increasingly rapidly until quality becomes zero
export default class TimeSensitive extends ItemDecorator {
  static MAXIMUM_QUALITY = 50;

  tick() {
    this.daysRemaining -= 1;

    if (this.daysRemaining < 0) {
      this.quality = 0;
      return;
    }

    this.quality += this.#calculateIncrease();
    this.quality = Math.min(this.quality, this.constructor.MAXIMUM_QUALITY);
  }

  #calculateIncrease() {
    let increase = 1;

    if (this.daysRemaining < 10) {
      increase += 1;
    }
    if (this.daysRemaining < 5) {
      increase += 1;
    }

    return increase;
  }
}
